c/tskit/tables.c

```diff
@@ -5388,6 +5381,638 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+// }
+
+static int
+ibd_finder_find_sample_pair_index2(
+    tsk_ibd_finder_t *self, tsk_id_t sample0, tsk_id_t sample1)
+{
+    int i = 0;
+    while (i < (int) self->num_pairs) {
+        if ((((int) self->pairs[2 * i] == (int) sample0)
+                && ((int) self->pairs[2 * i + 1] == (int) sample1))
+            || (((int) self->pairs[2 * i] == (int) sample1)
+                   && ((int) self->pairs[2 * i + 1] == (int) sample0))) {
+            break;
+        }
+        i++;
+    }
+    if (i < (int) self->num_pairs) {
+        return i;
+    } else {
+        return -1;
+    }
+}
+
+static int TSK_WARN_UNUSED
+ibd_finder_calculate_ibd(tsk_ibd_finder_t *self, tsk_id_t current_parent)
+{
+    int ret = 0;
+    int j, pair_index;
+    tsk_segment_t *seg, *seg0, *seg1;
+    double l, r;
+
+    if (self->ancestor_map_head[current_parent] == NULL) {
+        for (j = 0; j != (int) self->segment_queue_size; j++) {
+            seg = &self->segment_queue[j];
+            ret = ibd_finder_add_ancestry(
+                self, current_parent, seg->left, seg->right, seg->node);
+            if (ret != 0) {
+                goto out;
+            }
+        }
+    } else {
+        for (seg0 = self->ancestor_map_head[current_parent]; seg0 != NULL;
+             seg0 = seg0->next) {
+            for (j = 0; j != (int) self->segment_queue_size; j++) {
+                seg1 = &self->segment_queue[j];
+                if (seg0->node == seg1->node) {
+                    continue;
+                }
+                ret = ibd_finder_find_sample_pair_index2(self, seg0->node, seg1->node);
+                if (ret < 0) {
+                    continue;
+                }
+                pair_index = ret;
+
+                if (seg0->left > seg1->left) {
+                    l = seg0->left;
+                } else {
+                    l = seg1->left;
+                }
+                if (seg0->right < seg1->right) {
+                    r = seg0->right;
+                } else {
+                    r = seg0->right;
+                }
+
+                if (l < r) {
+                    if (r - l > self->min_length) {
+                        ret = ibd_finder_add_output(
+                            self, l, r, current_parent, pair_index);
+                        if (ret != 0) {
+                            goto out;
+                        }
+                    }
+                }
+            }
+        }
+        for (j = 0; j != (int) self->segment_queue_size; j++) {
+            seg = &self->segment_queue[j];
+            ret = ibd_finder_add_ancestry(
+                self, current_parent, seg->left, seg->right, seg->node);
+            if (ret != 0) {
+                goto out;
+            }
+        }
+    }
+    self->segment_queue_size = 0;
+
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_ibd_finder_get_ibd_segments(
+    tsk_ibd_finder_t *self, tsk_id_t pair_index, tsk_segment_t *ret_ibd_segments_head)
```

Comment:
I think it would be better do this slightly differently:
```c
int TSK_WARN_UNUSED
￼tsk_ibd_finder_get_ibd_segments(
￼    tsk_ibd_finder_t *self, tsk_id_t pair_index, tsk_segment_t **ret_ibd_segments_head)
{
     // error checking as before 
     *ret_ibd_segments_head = self->ibd_segments_head[pair_index]
}
```
Notice that we're passing in a pointer-to-a-pointer here. Client code would then look like

```c
segment_t *seg;

ret = tsk_ibd_finder_get_ibd_segments(ibd_finder, 0, &seg);
if (ret..) // .. blah blah

while (seg != NULL) {
     // process IBD segment
    seg = seg->next;
}
```